// windinsky/lib/windinsky.js
/*
	author: wangningbo
	data:2013/11/17
	modify-log:
	-----------------------------------------------------------------------------------------
	|	NAME 		DATE		LOG 														|
	-----------------------------------------------------------------------------------------
*/


var utils = require('../utils');
var windinsky = {
	version: '1.0.1',
	author: 'windinsky-company',
	router: require('router'),
	middleware: require('middleware'),
	controller: require('controller'),
	config:{
		'views' : '/views',
		'controllers' : '/controllers'
	}
};

var set = exports.set = function(key,value){
	if (key === 'api') {
		windinsky.router.setApiPath(value);
	};
	if (key === 'views') {
		windinsky.middleware.set('view',value);
	};
	if (key === 'views options') {
		windinsky.middleware.set('view options',value);
	};
	if (key === 'view engine') {
		windinsky.middleware.setViewEngine(value);
	};
	windinsky.config[key] = value || windinsky.config[key];
};
set('view engine',require('ejs'));

exports.use = function(module_name, module){
	windinsky[util.toCamule(module_name)] = module || require(module_name);
};

exports.add = windinsky.middleware.add;

exports.process = function(req,res){
	check_cfg();
	var req = windinsky.middleware.wrapRequest(req),
		res = windinsky.middleware.wrapResponse(res);
	windinsky.router.route(windinsky.config.controllers,windinsky.config.views,req,res);
};

exports.controller = windinsky.controller;
exports.utils = utils;

function check_cfg(){
	if (!windinsky.config['view engine']) {
		throw '================================================================='+
			'\nMiss view engine.'+
			'\nplease add:'+
				'\n\twindinsky.use("views engine",your_view_engine_module);'+
			'\nto your start-server-script.'+
			'\nFor example:'+
			'\n\twindinsky.use("views engine",require("ejs"));'+
			'\n=================================================================';
	};
	if (!windinsky.config['views']) {
		throw '================================================================='+
			'\nMiss views path config.'+
			'\nplease add:'+
				'\n\twindinsky.set("views","your_views_folder_path");'+
			'\nto your start-server-script.'+
			'\n=================================================================';
	};
}
